pkg/controller/nodeipam/ipam/multi_cidr_priority_queue.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ipam
+
+import (
+	"math"
+
+	cidrset "k8s.io/kubernetes/pkg/controller/nodeipam/ipam/multicidrset"
+)
+
+// A PriorityQueue implementation based on https://pkg.go.dev/container/heap#example-package-PriorityQueue
+
+// An PriorityQueueItem is something we manage in a priority queue.
+type PriorityQueueItem struct {
+	clusterCIDR *cidrset.ClusterCIDR
+	// labelMatchCount is the first determinant of priority.
+	labelMatchCount int
+	// selectorString is a string representation of the labelSelector associated with the cidrSet.
+	selectorString string
+	// index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
+}
+
+// A PriorityQueue implements heap.Interface and holds PriorityQueueItems.
+type PriorityQueue []*PriorityQueueItem
+
+func (pq PriorityQueue) Len() int { return len(pq) }
+
+// Less compares the priority queue items, to store in a min heap.
+// Less(i,j) == true denotes i has higher priority than j.
+func (pq PriorityQueue) Less(i, j int) bool {
+	if pq[i].labelMatchCount != pq[j].labelMatchCount {
+		// P0: CidrSet with higher number of matching labels has the highest priority.
+		return pq[i].labelMatchCount > pq[j].labelMatchCount
+	}
+
+	// If the count of matching labels is equal, compare the max allocatable pod CIDRs.
+	if pq[i].maxAllocatable() != pq[j].maxAllocatable() {
+		// P1: CidrSet with fewer allocatable pod CIDRs has higher priority.
+		return pq[i].maxAllocatable() < pq[j].maxAllocatable()
+	}
+
+	// If the value of allocatable pod CIDRs is equal, compare the node mask size.
+	if pq[i].nodeMaskSize() != pq[j].nodeMaskSize() {
+		// P2: CidrSet with a PerNodeMaskSize having fewer IPs has higher priority.
+		// For example, `27` (32 IPs) picked before `25` (128 IPs).
+		return pq[i].nodeMaskSize() > pq[j].nodeMaskSize()
+	}
+
+	// If the per node mask size are equal compare the CIDR labels.
+	if pq[i].selectorString != pq[j].selectorString {
+		// P3: CidrSet having label with lower alphanumeric value has higher priority.
+		return pq[i].selectorString < pq[j].selectorString
+	}
+
+	// P4: CidrSet having a smaller IP address value has a higher priority.
+	return pq[i].cidrLabel() < pq[j].cidrLabel()
```

Comment:
string comparison doesn't work , at least for IPv6, you can use the new netip package or convert to bigint to compare them https://go.dev/play/p/gKJTpJvuBzM

Reply:
Modified the comment to say `alpha-numerically smaller..`, this is the last check to avoid arbitrary selection,